### PR TITLE
Apply HaveShieldedRequirements to coinbase transactions

### PR DESCRIPTION
**Pull request by str4d**
Both transparent and shielded inputs have contextual checks that need to be enforced in the consensus rules. For shielded inputs, these are that the anchors in transactions correspond to real commitment tree states (to ensure that the spent notes existed), and that their nullifiers are not being double-spent.

When Sprout was first added to the codebase, we added input checks in the same places that transparent inputs were checked (https://github.com/zcash/zcash/pull/625/commits/a8ac403db0bde0c202c87646d3d40efcba788d59); namely anywhere `CCoinsViewCache::HaveInputs` is called. These all happened to be gated on `!tx.IsCoinBase()`, which was fine because we did not allow Sprout JoinSplits in coinbase transactions (enforced with a non-contextual check).

When we added Sapling we also allowed coinbase outputs to Sapling addresses (shielded coinbase). We updated `HaveShieldedRequirements` to check Sapling anchors and nullifiers, but didn't change the consensus code to call it on coinbase. This was fine because Sapling Spends and Outputs are separate, and we did not allow Sapling Spends in coinbase transactions (meaning that there were no anchors or nullifiers to enforce the input rules on).

Orchard falls into an interesting middle-ground:
- We allowed coinbase outputs to Orchard addresses, to enable Sapling shielded coinbase users to migrate to Orchard.
- Orchard uses Actions, which are a hybrid of Sprout JoinSplits and Sapling Spends/Outputs. That is, an Orchard Action comprises a single spend and a single output.

To maintain the "no shielded spends in coinbase" rule, we added an `enableSpends` flag to the Orchard circuit. We force it to be set to `false` for coinbase, ensuring that all Orchard spends in a coinbase use dummy (zero-valued) notes. However, this is insufficient: the coinbase transaction will still contain an Orchard anchor and nullifiers, and these need to be correctly constrained.

In particular, not constraining the Orchard nullifiers in a coinbase transaction enables a Faerie Gold attack. We explicitly require that Orchard nullifiers are unique, so that there is a unique input to the nullifier derivation. Without the coinbase check, the following attack is possible:
- An adversary creates an Orchard Action sending some amount of ZEC to a victim address, with a dummy spent note. The entire transaction can be fully-shielded by placing the real spent note in a separate Action.
- The adversary uses the exact same dummy note in a coinbase transaction, creating the exact same output note (same victim address and amount).
- The victim now has two notes with the same ZEC amount, but can only spend one of them because they have the same nullifier.

This was an implementation rather than a specification bug: the protocol specification already required that *all* Orchard nullifiers are unique (in [§ 3.9 'Nullifier Sets'](https://zips.z.cash/protocol/protocol.pdf#nullifierset)), and that *all* Orchard anchors reference the commitment tree root for the output treestate of a previous block (in [§ 4.6 'Action Descriptions'](https://zips.z.cash/protocol/protocol.pdf#actiondesc)). So as well as the attack described above, there was the potential for consensus divergence with implementations that correctly followed the spec.

This PR fixes the consensus bug by calling `HaveShieldedRequirements` outside of `!tx.IsCoinBase()` gates. To simplify its usage, there is now a `Consensus::CheckTxShieldedInputs` function that handles the logging and validation state updates. We also move shielded input checks from `ContextualCheckInputs` to `ContextualCheckShieldedInputs`; these now mirror each other in that they check contextual rules on transparent and shielded inputs respectively, followed by checking signatures.

### CODE REVIEW

**Review by ebfull — APPROVED** [review]

**Review by daira — APPROVED** [review]
utACK

File: src/txmempool.cpp [hunk 1]
```diff
@@ -587,38 +587,13 @@ void CTxMemPool::check(const CCoinsViewCache *pcoins) const
         // with randomness generated on construction.
         boost::unordered_map<uint256, SproutMerkleTree, SaltedTxidHasher> intermediates;
```
Comment (therealyingtong) [inline src/txmempool.cpp:588]: I don't think `intermediates` is being used anywhere now.

```suggestion
```

**Review by therealyingtong — APPROVED** [review]
utACK with a suggestion.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash/compare/934053980240dfa6ee8b9a49acf9d654a1d8624f..670800bd61b671c74e56ace81a20cf5732371cb0) to address @therealyingtong's comment, and update tests for the change to the `ContextualCheckShieldedInputs` method.

**Review by ebfull — APPROVED** [review]

**Review by therealyingtong — APPROVED** [review]
utACK 670800bd61b671c74e56ace81a20cf5732371cb0

**Review by daira — APPROVED** [review]
utACK

**Review by nuttycom — CHANGES_REQUESTED** [review]
Adding this block until CI starts working.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash/compare/670800bd61b671c74e56ace81a20cf5732371cb0..bb072f06c9f3e888da19a8b3da98ca847c3a165e) to fix the return value of the `AssumeShieldedInputsExistAndAreSpendable` test fixture (it was treating all shielded inputs as spent, rather than spendable as intended).

**Review by daira — APPROVED** [review]
utACK bb072f06c9f3e888da19a8b3da98ca847c3a165e

**Review by nuttycom — APPROVED** [review]
utACK https://github.com/zcash/zcash/commit/bb072f06c9f3e888da19a8b3da98ca847c3a165e

**Review by ebfull — APPROVED** [review]

**Pull request merged**